src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+            auto waiter = consensus->visible_offset_monitor()
+                            .wait(offset, model::no_timeout, _as)
+                            // All exceptions are ignored here as this is only
+                            // used to signal the worker that another attempt to
+                            // read the partition should be made.
+                            .handle_exception([](const std::exception_ptr&) {});
+
+            ssx::spawn_with_gate(
+              _waiter_gate, [this, w = std::move(waiter), i]() mutable {
+                  return w.finally([this, i] {
+                      _request_indexes.push_back(i);
+                      _has_completed_waiters.signal();
+                  });
+              });
+        }
+
+        return {};
+    }
+
+    ss::future<worker_result> do_run() {
+        bool first_run{true};
+        bool has_error{false};
+        // A map of indexes in `requests` to their corresponding index in
+        // `_ctx.requests`.
+        std::vector<size_t> requests_map;
+
+        std::vector<read_result> results;
+        size_t total_size{0};
+
+        for (;;) {
+            std::vector<ntp_fetch_config> requests;
+
+            if (first_run) {
+                requests = _ctx.requests;
+            } else {
+                requests_map.clear();
+
+                for (auto i : _request_indexes) {
+                    requests.push_back(_ctx.requests[i]);
+                    requests_map.push_back(i);
+                }
+
+                _request_indexes.clear();
+            }
+
+            auto q_results = co_await query_requests(std::move(requests));
+            has_error = q_results.has_error;
+            if (first_run) {
+                results = std::move(q_results.results);
+                total_size = q_results.total_size;
+
+                _last_visible_indexes = std::move(
+                  q_results.last_visible_indexes);
+            } else {
+                // Override the older results of the partitions with the newly
+                // queried results.
+                for (size_t i = 0; i < requests_map.size(); i++) {
+                    auto r_i = requests_map[i];
+                    auto& r = results[r_i];
+                    total_size -= r.data_size_bytes();
+                    r = std::move(q_results.results[i]);
+                    total_size += r.data_size_bytes();
+
+                    _last_visible_indexes[r_i]
+                      = q_results.last_visible_indexes[i];
+                }
+            }
+
+            if (
+              total_size >= _ctx.min_bytes || has_error
+              || _as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            std::optional<size_t> has_errored_request;
+            if (first_run) {
+                has_errored_request = register_waiters(
```

Comment:
Making register_waiter a template and passing it two differnet types is way overkill here I think. It would be better to just always pass requests_map and above populate it with all indexes on the first pass.

Reply:
It'd involve allocating enough space for potentially 2,000 elements even though and subsequent uses of the container will have 1 or 2 elements most of the time.

Reply:
I see, this is a good point, I like this approach.